packages/@lwc/integration-types/src/decorators/wire.ts

```diff
@@ -0,0 +1,142 @@
+    // @ts-expect-error prop must be initialized or set as optional
+    @wire(FakeWireAdapter, { config: 'config' }) nonUndefinedProp: WireValue;
+    // @ts-expect-error `wrongProp` is not a valid reactive prop
+    @wire(FakeWireAdapter, { config: '$wrongProp' } as const) nonexistentReactiveProp?: WireValue;
+    // @ts-expect-error `nested.wrong` is not a valid reactive prop
+    @wire(FakeWireAdapter, { config: '$nested.wrong' } as const)
+    wrongNestedReactiveProp?: WireValue;
+    // @ts-expect-error `otherProp` is the wrong type
+    @wire(FakeWireAdapter, { config: '$otherProp' } as const) wrongReactivePropType?: WireValue;
+    // @ts-expect-error nested props are not reactive
+    @wire(DeepConfigAdapter, { deep: { config: '$otherProp' } })
+    nonReactiveDeepObjectConfig?: WireValue;
+
+    // Ambiguous case: Passing a config is optional because adapters don't strictly need to use it.
+    // Can we be smarter about the type and require a config if the adapter does?
+    @wire(FakeWireAdapter) noConfigProp?: WireValue;
+}
+
+/** Validations for decorated methods */
+export class MethodDecorators extends LightningElement {
+    plainProp = 'config' as const;
+    otherProp = 123;
+    nested = { object: 'config' as const };
+    // Valid cases
+    @wire(FakeWireAdapter, { config: 'config' }) baseConfig(_: WireValue) {}
+    @wire(FakeWireAdapter, { config: 'config' }) optionalParam(_?: WireValue) {}
+    @wire(FakeWireAdapter, config) configAsVar(_: WireValue) {}
+    @wire(FakeWireAdapter, { config: '$plainProp' }) reactiveConfig(_: WireValue) {}
+    @wire(FakeWireAdapter, { config: '$nested.object' }) nestedReactiveConfig(_: WireValue) {}
+    @wire(FakeWireAdapter, { config: 'config' } as const) baseConstConfig(_: WireValue) {}
+    @wire(FakeWireAdapter, { config: '$plainProp' } as const) reactiveConstConfig(_: WireValue) {}
+    @wire(FakeWireAdapter, { config: '$nested.object' } as const)
+    nestedReactiveConstConfig(_: WireValue) {}
+    @wire(FakeWireAdapter, { config: 'config' }) emptyMethod() {}
+
+    // Invalid cases -- method
+    // @ts-expect-error prop type is `string` but the adapter needs `WireValue`
+    @wire(FakeWireAdapter, { config: 'config' }) wrongMethodSignature(_: 'wrong type') {}
+    // @ts-expect-error config type is `{wrong: string}` but the adapter needs `WireConfig`
+    @wire(FakeWireAdapter, { wrong: 'type' }) wrongConfigType(_: WireValue) {}
+    // @ts-expect-error too many arguments
+    @wire(FakeWireAdapter, { config: 'config' })
+    tooManyArguments(_a: WireValue | undefined, _b: unknown): void {}
+
+    // Ambiguous case: Passing a config is optional because adapters don't strictly need to use it.
+    // Can we be smarter about the type and require a config if the adapter does?
+    @wire(FakeWireAdapter) noConfig(_: WireValue) {}
+
+    // These types are inferred as `string`, so we can't do any further checking on them :\
```

Comment:
Why this limitation?

Reply:
```ts
class extends LightningElement {
  id = 123
  @wire(getThing, {id: '$id'}) valid;
  @wire(getThing, {id: '123'}) invalid;
}
```
Here, the `getThing` adapter expects a config with the type `{id: number}`. But, in both cases, the config provided is of type `{id: string}`. We can't distinguish a valid `string` (the first case) from an invalid one (the second), so we allow them both.

We technically could be strict; if users add `as const` to all of their wire decorators (`@wire(getThing, {id: '$id'} as const)`), we could get accurate type checking. In the above example, that would result in type `{id: '$id'}`, which we can map to `{id: number}`, and type `{id: '123'}`, which we can see is invalid. However, if we enforce this and you _don't_ use `as const` in your decorators, then the error is not necessarily clear, and the path to resolution less so. And given that approximately nobody is currently using `as const` in their wire decorators (there's currently no reason to do so), then we'd be inflicting pain on everybody for very marginal gain.

Reply:
Makes sense, thanks!